src/common/lib/client/realtimepresence.ts

```diff
@@ -104,41 +104,56 @@ class RealtimePresence extends Presence {
     if (isAnonymousOrWildcard(this)) {
       throw new ErrorInfo('clientId must be specified to enter a presence channel', 40012, 400);
     }
-    return this._enterOrUpdateClient(undefined, undefined, data, 'enter', callback);
+    if (typeof data === 'function') {
+      callback = data as ErrCallback;
+      data = null;
+    }
+    const msg = PresenceMessage.fromData(data);
+    msg.action = 'enter';
```

Comment:
could we just move these 5 lines to `_enterOrUpdateClient` and take `msg.action` as an argument rather than repeating it 4 times?

Reply:
Ah yes, I've moved the `fromData` call into `_enterOrUpdateClient` which removes the duplication (and reduces the diff), see 5baf94d9